### PR TITLE
Bashisms

### PR DESCRIPTION
**Pull request by scopatz**
This starts a bashisms xontrib that performs the mininal transformation.  In order to get it working, some refactoring and cleanup was needed.

### CODE REVIEW

**Comment by scopatz** [comment]
ping!

**Comment by melund** [comment]
I haven't had a chance to try it but I did have some time to read the code on my phone. It looks good to me. The precommand event is quite interesting. I think it will be extremely useful for a lot of application specific things.

Unless you want to give others a chance to test I am OK with merging this in.

**Comment by gforsyth** [comment]
I was traveling yesterday but I'll take a look at this today

File: xonsh/shell.py [hunk 1]
```diff
@@ -27,6 +29,34 @@
...
+        lst = src
+        srcs = events.on_precommand.fire(src)
+        for s in srcs:
+            if s != lst:
+                src = s
+                break
+        i += 1
+        if i == limit:
+            print_exception('Modifcations to source input took more than '
+                            'the recurssion limit number of interations to '
```
Comment (gforsyth) [inline xonsh/shell.py]: recursion

Comment (scopatz) [inline xonsh/shell.py]: Got it, thanks.

File: xonsh/xontribs.json [hunk 2]
```diff
@@ -1,4 +1,25 @@
...
+ },
+ {"name": "bashisms",
+  "package": "xonsh",
+  "url": "http://xon.sh",
+  "description": [
+    "Enables additional Bash-like syntax while at the command prompt. For ",
+    "example, the ``!!`` syntax for running the previous command is now usable.",
+    "Note that these features are implemented as precommand events and these ",
+    "additions do not affect the xonsh language when run as script. That said, ",
+    "you might find them useful if you have strong musucle memory."]
```
Comment (gforsyth) [inline xonsh/xontribs.json]: muscle

**Comment by gforsyth** [comment]
Ok, this works and I like it.  

Thought and question: I had been working on something similar and while there is no way around it, transforming input before it's executed would seem to open `xonsh` up to malicious xontribs.  Should there be a warning the first time that a xontrib is going to modify what was entered by the user before execution?

**Comment by gforsyth** [comment]
Also, to be fair, there could also be a xontrib that just does `rm -rf /` whenever you load it and that would have nothing to do with transforming `src`

**Comment by scopatz** [comment]
Also, this is pretty simple and the actual implementation could be really expanded.

**Comment by scopatz** [comment]
The test failure might be related to circle's github service being down.  @laerus, do you have any ideas here since I think this it related to your recent VC work.  Is it OK to not do the assert if we get back a None?

**Comment by gforsyth** [comment]
I think we can also take the time to review xontribs that are submitted to make sure they aren't malicious.  Just waiting on CI for this, then

**Comment by gforsyth** [comment]
The circle failure seems to have been a fluke...?

**Comment by scopatz** [comment]
> The circle failure seems to have been a fluke...?

I added the workaround :)

**Comment by scopatz** [comment]
CI finally finished :)

**Comment by gforsyth** [comment]
Awesome!  Thanks @scopatz!

**Pull request merged**